core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+     * @param inputPage A page containing write redistribution column blocks for the target table; the
+     * rowId block; the merge case RowBlock; and for partitioned or bucketed tables, a hash column block.
+     * @return A page containing all data columns, the operation block and the rowId block,
+     * with UPDATE rows expanded into delete and insert rows, so they can be routed separately.
+     * The delete rows contain the partition key blocks from the target table, whereas the insert
+     * rows have the partition key blocks from the merge case RowBlock.
+     */
+    @Override
+    public Page transformPage(Page inputPage)
+    {
+        requireNonNull(inputPage, "inputPage is null");
+        int inputChannelCount = inputPage.getChannelCount();
+        if (inputChannelCount < 2 + redistributionColumnCount) {
+            throw new IllegalArgumentException(format("inputPage channelCount (%s) should be >= 2 + partition columns size (%s)", inputChannelCount, redistributionColumnCount));
+        }
+
+        int originalPositionCount = inputPage.getPositionCount();
+        if (originalPositionCount <= 0) {
+            throw new IllegalArgumentException("originalPositionCount should be > 0, but is " + originalPositionCount);
+        }
+        Block mergeCaseBlock = inputPage.getBlock(redistributionColumnCount + 1);
+        List<Block> mergeCaseBlocks = mergeCaseBlock.getChildren();
+        int mergeBlocksSize = mergeCaseBlocks.size();
+        Block rowMatchedBlock = mergeCaseBlocks.get(mergeBlocksSize - 2);
+        Block operationChannelBlock = mergeCaseBlocks.get(mergeBlocksSize - 1);
+        duplicateRowFinder.checkForDuplicateTargetRows(inputPage, operationChannelBlock);
+
+        int updatePositions = 0;
+        int insertPositions = 0;
+        int deletePositions = 0;
+        for (int position = 0; position < originalPositionCount; position++) {
+            int operation = operationChannelBlock.getInt(position, 0);
+            switch (operation) {
+                case DEFAULT_CASE_OPERATION_NUMBER:
+                    break;
+                case INSERT_OPERATION_NUMBER:
+                    insertPositions++;
+                    break;
+                case DELETE_OPERATION_NUMBER:
+                    deletePositions++;
+                    break;
+                case UPDATE_OPERATION_NUMBER:
+                    updatePositions++;
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown operator number " + operation);
+            }
+        }
+
+        int totalPositions = insertPositions + deletePositions + 2 * updatePositions;
+        Block rowIdBlock = inputPage.getBlock(redistributionColumnCount);
+
+        ImmutableList.Builder<Type> pageTypesBuilder = ImmutableList.builder();
+        pageTypesBuilder.addAll(dataColumnTypes);
+        pageTypesBuilder.add(INTEGER);
```

Comment:
Should this be `TINYINT` since we only have a few distinct values?

Reply:
It would need to get changed in a bunch of different places.  I'll do it as a follow-up.

Reply:
This would affect SPI, so it's better to do it first iteration.

Reply:
I made this change - - the operation column is now a TINYINT.